ranger/config/commands.py

```diff
@@ -1176,24 +1177,86 @@ def execute(self):
 
 
 class chmod(Command):
-    """:chmod <octal number>
+    """:chmod [<octal number>|<mode-spec>]
 
-    Sets the permissions of the selection to the octal number.
+    Sets the permissions of the selection to the given mode.
 
     The octal number is between 0 and 777. The digits specify the
     permissions for the user, the group and others.
 
     A 1 permits execution, a 2 permits writing, a 4 permits reading.
     Add those numbers to combine them. So a 7 permits everything.
+
+    The <mode-spec> is a comma-separated where each entry has the
+    form [ugoa]+[+-][rwx]+
     """
 
-    def execute(self):
-        mode_str = self.rest(1)
+    def __init__(self, *args, **kwargs):
+        '''init chmod-command.'''
+        super(chmod, self).__init__(*args, **kwargs)
+        self.re_mode = None
+        self.bitmasks = None
+
+    def prepare(self):
+        '''create regex for mode-spec and bitmasks.'''
+        self.re_mode = re.compile(r'''
+            ([ugoa]+)     # (u)ser, (g)roup, (o)other, (a)ll
+            ([+-])        # add (+), remove (-)
+            ([rwx]+)      # (r)ead, (w)rite, e(x)ecute
+            $             # end
+            ''', re.X)
+        masks = {
```

Comment:
Good start and looks equivalent to the octal implementation but let's try expanding this to cover the full grammar.